mycroft/tts/tts.py

```diff
@@ -351,22 +355,46 @@ def _execute(self, sentence, ident, listen):
                   for i in range(len(chunks))]
 
         for sentence, l in chunks:
-            key = str(hashlib.md5(
-                sentence.encode('utf-8', 'ignore')).hexdigest())
-            wav_file = os.path.join(
-                mycroft.util.get_cache_directory("tts/" + self.tts_name),
-                key + '.' + self.audio_ext)
-
-            if os.path.exists(wav_file):
-                LOG.debug("TTS cache hit")
-                phonemes = self.load_phonemes(key)
+            sentence_hash = hash_sentence(sentence)
+            if sentence_hash in self.cache.cached_sentences:
+                audio_file, phoneme_file = self._get_sentence_from_cache(
+                    sentence_hash
+                )
+                if phoneme_file is None:
+                    phonemes = None
+                else:
+                    phonemes = phoneme_file.load()
+
             else:
-                wav_file, phonemes = self.get_tts(sentence, wav_file)
+                # TODO: this should be changed return the audio data from
+                #  the API call and then to call the add_to_cache method
+                #  of the TTS cache.  But this requires changing the public
+                #  API of the get_tts method in each engine.
+                audio_file, phoneme_file = self._add_sentence_to_cache(
+                    sentence_hash
+                )
+                _, phonemes = self.get_tts(sentence, str(audio_file.path))
```

Comment:
Isn't it clearer to rely on the return than use the side-effect?

Is the order ideal here.

First we add something to the cache then we try to get / generate the content of that cache. What would happen if the fetch/generation fails. Would there be a cache entry with no data?